### PR TITLE
[install] support trustedDependencies

### LINKED ISSUE #2073: Support `trustedDependencies` in `package.json`
There are npm packages, like `esbuild`, that need `postinstall` support to install binaries or run other scripts. While we could allow any package to run scripts, that wouldn't be the safest thing to do. We also can't _not_ support it at all, because these packages would simply not work.

We're thinking of adding a new `trustedDependencies` property to the `package.json`, which would require developers to explicitly define an allowlist of packages that are allowed to run scripts.

```json
{
  "private": true,
  "dependencies": {
    "esbuild": "^0.17.8"
  },
  "trustedDependencies": [
    "esbuild"
  ]
}
```

This would not extend to child dependencies, so if package `foo` depends on `bar`, and `foo` is on the allowlist, `bar` is _not_ on the allowlist, unless it is explicitly added.

### PR DESCRIPTION
**Pull request by alexlamsl**
closes #2073

### CODE REVIEW

**Comment by Electroid** [comment]
Note: we're gonna rename it to `trustedDependencies`

File: src/install/extract_tarball.zig [hunk 1]
```diff
@@ -157,13 +157,14 @@ fn extract(this: *const ExtractTarball, tgz_bytes: []const u8) !Install.ExtractD
...
     var tmpname_buf: [256]u8 = undefined;
     const name = this.name.slice();
-
-    var basename = this.name.slice();
-    if (basename[0] == '@') {
-        if (std.mem.indexOfScalar(u8, basename, '/')) |i| {
-            basename = basename[i + 1 ..];
+    const basename = brk: {
+        if (name[0] == '@') {
+            if (std.mem.indexOfScalar(u8, name, '/')) |i| {
```
Comment (Jarred-Sumner) [inline src/install/extract_tarball.zig]: nit: `strings.indexOfChar(name, '/')` is faster (though it shouldn't matter here)

File: src/install/lockfile.zig [hunk 2]
```diff
@@ -3069,6 +3073,37 @@ pub const Package = extern struct {
             }
         }
 
+        if (comptime features.privileged_dependencies) {
+            if (json.asProperty("privilegedDependencies")) |q| {
+                switch (q.expr.data) {
+                    .e_array => |arr| {
+                        try lockfile.privileged_dependencies.ensureUnusedCapacity(allocator, arr.items.len);
+                        for (arr.slice()) |item| {
```
Comment (Jarred-Sumner) [inline src/install/lockfile.zig:3124]: i'm sorry this is a really annoying suggestion but

could we count the total length here and then use StringBuilder? These might become a long list

File: src/install/extract_tarball.zig (hunk 1 above)
Comment (alexlamsl) [reply src/install/extract_tarball.zig]: Replaced all usages of `std.mem.indexOfScalar(u8, ...)` and `std.mem.lastIndexOfScalar(u8, ...)` within `src/install`

File: src/install/lockfile.zig (hunk 2 above)
Comment (alexlamsl) [reply src/install/lockfile.zig:3124]: Dropped strings altogether &minus; store and compares hashes instead 😉

**Comment by arcanis** [comment]
FYI Yarn already has a syntax in the `package.json` for this; docs [here](https://yarnpkg.com/configuration/manifest#dependenciesMeta.built):

```json
{
  "dependenciesMeta": {
    "some-dep": {
      "built": true
    }
  }
}
```

**Pull request merged**